cairo/ethereum/cancun/vm/memory.cairo

```diff
@@ -0,0 +1,202 @@
+    }
+    let new_len = start_position_felt + bytes_len;
+
+    let dict_ptr = cast(memory.value.dict_ptr, DictAccess*);
+    with dict_ptr {
+        Internals._write_bytes(start_position_felt, bytes_data, bytes_len);
+    }
+    let new_dict_ptr = cast(dict_ptr, Bytes1DictAccess*);
+
+    // Update length if we wrote beyond current length
+    tempvar current_len = memory.value.len;
+    let is_new_le_current = is_le(new_len, current_len);
+    if (is_new_le_current != TRUE) {
+        tempvar final_len = new_len;
+    } else {
+        tempvar final_len = current_len;
+    }
+
+    tempvar memory = Bytearray(
+        new BytearrayStruct(memory.value.dict_ptr_start, new_dict_ptr, final_len)
+    );
+    return ();
+}
+
+// @notice Read bytes from memory.
+// @dev assumption: start_position < 2**128
+// @dev assumption: size < 2**128
+// @param memory The pointer to the bytearray.
+// @param start_position Starting position to read from.
+// @param size Number of bytes to read.
+// @return The bytes read from memory.
+func memory_read_bytes{memory: Bytearray}(start_position: U256, size: U256) -> Bytes {
+    alloc_locals;
+    let (local output: felt*) = alloc();
+
+    let start_position_felt = start_position.value.low;
+    let size_felt = size.value.low;
+    with_attr error_message("memory_read_bytes: start_position > 2**128 || size > 2**128") {
+        assert start_position.value.high = 0;
+        assert size.value.high = 0;
+    }
+    let dict_ptr = cast(memory.value.dict_ptr, DictAccess*);
+    with dict_ptr {
+        Internals._read_bytes(start_position_felt, size_felt, output);
+    }
+    let new_dict_ptr = cast(dict_ptr, Bytes1DictAccess*);
+
+    tempvar memory = Bytearray(
+        new BytearrayStruct(memory.value.dict_ptr_start, new_dict_ptr, memory.value.len)
+    );
+    tempvar result = Bytes(new BytesStruct(output, size_felt));
+    return result;
+}
+
+// @notice Read bytes from a buffer with zero padding.
+// @dev assumption: start_position < 2**128
+// @dev assumption: size < 2**128
+// @param buffer Source bytes to read from.
+// @param start_position Starting position to read from.
+// @param size Number of bytes to read.
+// @return The bytes read from the buffer.
+func buffer_read{range_check_ptr}(buffer: Bytes, start_position: U256, size: U256) -> Bytes {
+    alloc_locals;
+    let (local output: felt*) = alloc();
+    let buffer_len = buffer.value.len;
+    let buffer_data = buffer.value.data;
+    let start_position_felt = start_position.value.low;
+    let size_felt = size.value.low;
+    with_attr error_message("buffer_read: start_position > 2**128 || size > 2**128") {
+        assert start_position.value.high = 0;
+        assert size.value.high = 0;
+    }
+
+    Internals._buffer_read(buffer_len, buffer_data, start_position_felt, size_felt, output);
+    tempvar result = Bytes(new BytesStruct(output, size_felt));
+    return result;
+}
+
+namespace Internals {
```

Comment:
let's not use namespace as
1. it doesn't exist in python, function prefixed with `_` are automatically discarded from `*` import
2. it doesn't work with the current test runner, see #220